### PR TITLE
add offline retry

### PR DESCRIPTION
**Pull request by markwolff**
retries on `xhr.status === 0` or `navigator.onLine === false`

### CODE REVIEW

File: Sender.ts [hunk 1]
```diff
@@ -233,6 +233,16 @@ export class Sender implements IChannelControlsAI {
                 } else {
                     this._onError(payload, this._formatErrorMessageXhr(xhr));
                 }
+            } else if (xhr.status === 0 || navigator.onLine === false) { // offline
```
Comment (jpiyali) [inline Sender.ts]: Any other conditions in other browsers that indicates offline?

File: Sender.ts [hunk 2]
```diff
@@ -233,6 +233,16 @@ export class Sender implements IChannelControlsAI {
...
+                if (!this._config.isRetryDisabled()) {
+                    const offlineBackOffMultiplier = 10; // arbritrary number
+                    this._resendPayload(payload, offlineBackOffMultiplier);
```
Comment (jpiyali) [inline Sender.ts]: for linear , we dont need to do it, but if we change this, we should look at pre-empting retry wait time if we can find a way to get notification for coming  back online.

File: Sender.ts [hunk 3]
```diff
@@ -233,15 +234,14 @@ export class Sender implements IChannelControlsAI {
                 } else {
                     this._onError(payload, this._formatErrorMessageXhr(xhr));
                 }
-            } else if (xhr.status === 0 || navigator.onLine === false) { // offline
+            } else if (xhr.status === 0 || Offline.isOffline()) { // offline
```
[the remark this answers is not in the document]
Comment (jpiyali) [reply Sender.ts]: I see the export at the bottom of other file now.

[the remark this answers is not in the document]
Comment (jpiyali) [reply Sender.ts]: Looks good.

[the remark this answers is not in the document]
Comment (markwolff) [reply Sender.ts]: yes its a static instance. For now, if multiple things need offline checking, we only ever need to instrument once

**Pull request merged**